keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,504 @@
+
+However, the problem raising by this KEP is that there is no hint provider which can give a hint for memory and hugepages resources. That is, Topology Manager could not determine NUMA-aware resource allocation in best way due to absence of memory-related information, which causes inter-node commucation overhead among CPU, memory, and PCIe devices and leads overall performance degradation of containers.
+
+To resolve the problem, Memory Manager is proposed in this KEP to provide a way to guarantee NUMA affinity of memory as well as other resources (i.e., CPU and PCIe peripherals) in a procedure of container deployment.
+
+# Motivation
+
+## Related Features
+
+- [Topology Manager][topology-manager] is a feature that collects topology hints from various hint providers (e.g., CPU Manager and Device Manager) to calculate which NUMA node can give a requested amount of resources for a container. With this calculated hints, Topology Manager also refers to node topology policy (i.e. _burstable, restricted, single-numa-policy_) and NUMA affinity of containers, and then it finally determines if container in a pod can be deployed onto the host machine.
+
+- [CPU Manager][cpu-manager] is a feature that provides a CPU pinning functionality to a container by cgroups cpuset subsystem and also provides topology hint of CPU core availability in NUMA nodes to Topology Manager.
+
+- [Device Manager][device-manager] is a feature that device vendors can advertise their device resources like NIC and GPU devices through their own device-plugins so that kubelet can discover the devices to be used by containers. And also, Device Manager can give a hint of corresponding device's availability in NUMA nodes to Topology Manager.
+
+- [Hugepages][hugepages] is a feature to assign pre-allocated hugepage resources to a container if requested.
+
+- [Node Allocatable Feature][node-allocatable-feature] is a feature that helps to reserve compute resources to isolate whole resources on a host by certain purpose. For example, by this feature, kube-reserved and system-reserved can be defined to reserve resources for kubelet and system (i.e. OS kernel), respectively. In release 1.17, the feature supports following reservable resources: CPU, memory, and ephemeral storage.
+
+## Related issues
+
+- [Hardware topology awareness at node level (including NUMA)][numa-issue]
+- [Support Container Isolation of Hugepages][hugepage-issue]
+- [Support isolating memory to single NUMA node][memory-issue]
+
+## Goals
+
+- To guarantee isolation of memory and hugepages to single NUMA node for containers belonging to the Guaranteed Pod.
```

Comment:
I believe we don't need to emphasise single NUMA allocations. It's just one of possible topology manager policies. We need to make this goal more generic, e.g. "To guarantee isolation of memory and hugepages to a minimal set of NUMA nodes" or something similar.

Reply:
Okay, at this point of document, it might be even more suitable to generalize, so I applied your suggestion.
But, I would opt to keep the terms single-NUMA and multi-NUMA for a variety of reasons (e.g. the Linux kernel does not support  multi-NUMA guaranteed memory allocation).